0099-Replacing bash as the default shell.md

```diff
@@ -0,0 +1,120 @@
+---
+feature: oil-shell
+start-date: 2021-08-11
+author: Raphael Megzari (happysalada)
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: Eelco Dolstra (edolstra), Kevin Cox (kevincox), Las Safin (L-as)
+shepherd-leader: Kevin Cox (kevincox)
+related-issues:
+---
+
+# Summary
+
+[summary]: #summary
+
+Use oil as the stdenv shell in nixos. Oil aims to be compatible with bash while trying to bring real programming language features (like hashmap or dicts) to the shell.
+
+## Motivation
+
+[motivation]: #motivation
+
+Two primary motivations
+
+- Removing the footguns in bash that even experienced programmers find painful
+- Making shell scripts more powerfull to remove the need to bring additional languages for scripting.
+
+## Detailed design
+
+[design]: #detailed-design
+
+The [oil shell](https://www.oilshell.org/) has two parts osh and the oil language. Osh is compatible with bash and posix and it's goal is to run existing shell scripts. The oil language is a brand new incompatible language. The idea is to fix more than four decades of accumulated warts in the Unix shell. Many Unix users are angry that shell is so difficult, and Oil aims to fix that. Those definitions were taken verbatim from [reference](https://www.oilshell.org/blog/2021/01/why-a-new-shell.html)
+
+Regarding oil
+
+A high level overview is that it has a syntax similar to python, it brings dictionaries (hashmaps). For more details check the following posts
+
+- [This post](https://www.oilshell.org/blog/2021/01/why-a-new-shell.html) aims at describing the goals in trying to create a new shell and the alternatives.
+- [This post](https://www.oilshell.org/blog/2020/01/simplest-explanation.html) aims to provide a simple explanation of what oil is.
+- [This post](https://www.oilshell.org/release/latest/doc/idioms.html) shows different oil idioms and how they fix some of bash problems.
+
+There would be two steps to make the transition from bash to oil.
+
+- From bash to osh. There is one incompatibility that came up so far. osh does not handle the `-i` flag on local and declare. The transition involves rewriting some shell scripts to get rid of the `-i` flag. There are about 5-10 uses in stdenv, and I've started PRs to remove them. After those changes, it would be possible to switch from bash to osh and test this for a while to verify there are no regression.
+- From osh to oil. It will require to modify stdenv to be able to run with the `strict_errexit` flag. The changes will be a little more involved, but as long as nobody uses oil specific features, we could retain compatibility with bash and posix. The exact quantity of work involved is unknown.
```

Comment:
Perhaps you can clarify the meaning of this statement, as the way it is currently written, it seems to defeat the purpose of this whole idea. Why rewrite the standard env if we don't actually wish to leverage any of oils specific features? If that's not what you meant to convey, then perhaps this segment should just be removed?

Reply:
What I mean by this is that the changes to make the switch possible are a little more involved.
The changes to osh only include removing a `-i` flag so far (potentially a couple of lines change in stdenv).
The changes to enable a switch to oil are more involved (hard to know the extent yet).
The referring to maintaining compatibility is related to having a period where things are in-between and we can evaluate if we want to switch to oil or not. Potentially retain compatibility for some time and when we feel more comfortable then start using oil exclusive feature.
I do agree with you, it's not very well formulated, how would the following work ?

```
- From osh to oil requires more heavy changes to stdenv. In particular it requires changing the logic so it can run with `-e` (bash equivalent, strict_errexit). After those changes are made, compatibility with bash can be retained for an arbitrary period of time until we are more confident we want to make the full switch (when we start using oil exclusive features).